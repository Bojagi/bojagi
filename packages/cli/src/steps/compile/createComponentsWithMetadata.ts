/* eslint-disable max-classes-per-file */
import { Module, ComponentWithMetadata } from '@bojagi/types';
import { ScannedComponent } from '../scan';

export class EntrypointError extends Error {}
export class FilePathError extends Error {}

export default function createComponentsWithMetadata(
  scanComponents: ScannedComponent[],
  compilerOutput: Record<string, string>,
  modules: Module[]
): ComponentWithMetadata[] {
  return scanComponents.map(scannedComponent => {
    const fileContent = compilerOutput[scannedComponent.fileName];

    const matchingModule = modules.find(module => module.filePath === scannedComponent.filePath);
    const dependencies = matchingModule && matchingModule.dependencies;
    return {
      ...scannedComponent,
      fileContent,
      dependencies,
    };
  });
}
