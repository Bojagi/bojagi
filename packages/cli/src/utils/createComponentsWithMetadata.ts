import { EntrypointWithMetadata } from '../commands/bundle';
import { Module } from '../types';

export class EntrypointError extends Error {}
export class FilePathError extends Error {}

export default function createComponentsWithMetadata(
  entrypointsWithMetadata: Record<string, EntrypointWithMetadata>,
  compilerOutput: Record<string, string>,
  modules: Module[]
) {
  return Object.entries(entrypointsWithMetadata).reduce((prev, [key, ep]) => {
    const fileContent = compilerOutput[key];
    if (!/^.+?\?.+?\!.+?$/.test(ep.entrypoint)) {
      throw new EntrypointError('Entrypoint does not follow right syntax');
    }

    const fullFilePath = ep.entrypoint.split('!')[1];
    const prefixPath = `${process.cwd()}/`;
    if (!fullFilePath.startsWith(prefixPath)) {
      throw new FilePathError('File does not start with correct path');
    }
    const filePath = fullFilePath.replace(new RegExp(`^${prefixPath}`), '');
    const matchingModule = modules.find(module => module.resource === filePath);
    const dependencies = matchingModule && matchingModule.dependencies;
    const components = ep.components.map(({ symbol, isDefaultExport }) => ({
      fileContent,
      symbol,
      isDefaultExport,
      filePath,
      dependencies,
      exportName: isDefaultExport ? 'default' : symbol,
      name: symbol
    }));
    return [...prev, ...components];
  }, []);
}
