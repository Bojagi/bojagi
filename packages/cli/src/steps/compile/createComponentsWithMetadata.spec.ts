import createComponentsWithMetadata from './createComponentsWithMetadata';

let scanComponents;
let compilerOutput;
let modules;

beforeEach(() => {
  scanComponents = [
    {
      fileName: 'MainComponent',
      filePath: 'mc-filePath',
      symbol: 'mc-c1-symbol',
      exportName: 'default',
      name: 'mc-c1-symbol',
      isDefaultExport: true,
    },
    {
      fileName: 'MainComponent',
      filePath: 'mc-filePath',
      exportName: 'mc-c2-symbol',
      symbol: 'mc-c2-symbol',
      name: 'mc-c2-symbol',
      isDefaultExport: false,
    },
    {
      fileName: 'SmallerComponent',
      filePath: 'sc-filePath',
      symbol: 'sc-c1-symbol',
      exportName: 'sc-c1-symbol',
      name: 'sc-c1-symbol',
      isDefaultExport: false,
    },
    {
      fileName: 'SmallerComponent',
      filePath: 'sc-filePath',
      symbol: 'sc-c2-symbol',
      exportName: 'default',
      name: 'sc-c2-symbol',
      isDefaultExport: true,
    },
  ];
  compilerOutput = {
    MainComponent: 'content of main component',
    SmallerComponent: 'content of smaller component',
  };
  modules = [
    {
      filePath: 'sc-filePath',
      dependencies: 'some deps',
    },
  ];
});

test('get components with metadata', () => {
  const result = createComponentsWithMetadata(scanComponents, compilerOutput, modules);
  expect(result).toEqual([
    {
      fileName: 'MainComponent',
      symbol: 'mc-c1-symbol',
      isDefaultExport: true,
      filePath: 'mc-filePath',
      fileContent: 'content of main component',
      exportName: 'default',
      name: 'mc-c1-symbol',
    },
    {
      fileName: 'MainComponent',
      symbol: 'mc-c2-symbol',
      isDefaultExport: false,
      filePath: 'mc-filePath',
      fileContent: 'content of main component',
      exportName: 'mc-c2-symbol',
      name: 'mc-c2-symbol',
    },
    {
      fileName: 'SmallerComponent',
      symbol: 'sc-c1-symbol',
      isDefaultExport: false,
      filePath: 'sc-filePath',
      fileContent: 'content of smaller component',
      exportName: 'sc-c1-symbol',
      name: 'sc-c1-symbol',
      dependencies: 'some deps',
    },
    {
      fileName: 'SmallerComponent',
      symbol: 'sc-c2-symbol',
      isDefaultExport: true,
      filePath: 'sc-filePath',
      fileContent: 'content of smaller component',
      exportName: 'default',
      name: 'sc-c2-symbol',
      dependencies: 'some deps',
    },
  ]);
});
