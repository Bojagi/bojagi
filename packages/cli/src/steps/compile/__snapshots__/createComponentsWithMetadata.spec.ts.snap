// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`fail to get components with metadata because entrypoint does not have correct syntax 1`] = `"Entrypoint does not follow right syntax"`;

exports[`fail to get components with metadata because file path doesn't start with current path 1`] = `"File does not start with correct path"`;
